content/guides/frameworks/laravel/development-setup.md

```diff
@@ -0,0 +1,342 @@
+    docker-php-ext-enable xdebug && \
+    echo "xdebug.mode=${XDEBUG_MODE}" >> /usr/local/etc/php/conf.d/docker-php-ext-xdebug.ini && \
+    echo "xdebug.idekey=${XDEBUG_IDE_KEY}" >> /usr/local/etc/php/conf.d/docker-php-ext-xdebug.ini && \
+    echo "xdebug.log=${XDEBUG_LOG}" >> /usr/local/etc/php/conf.d/docker-php-ext-xdebug.ini && \
+    echo "xdebug.log_level=${XDEBUG_LOG_LEVEL}" >> /usr/local/etc/php/conf.d/docker-php-ext-xdebug.ini && \
+    echo "xdebug.client_host=${XDEBUG_HOST}" >> /usr/local/etc/php/conf.d/docker-php-ext-xdebug.ini ; \
+    echo "xdebug.start_with_request=yes" >> /usr/local/etc/php/conf.d/docker-php-ext-xdebug.ini ; \
+fi
+
+# Set environment variables for user and group ID
+ARG UID=1000
+ARG GID=1000
+
+# Create a new user with the specified UID and GID, reusing an existing group if GID exists
+RUN if getent group ${GID}; then \
+      group_name=$(getent group ${GID} | cut -d: -f1); \
+      useradd -m -u ${UID} -g ${GID} -s /bin/bash www; \
+    else \
+      groupadd -g ${GID} www && \
+      useradd -m -u ${UID} -g www -s /bin/bash www; \
+      group_name=www; \
+    fi
+
+# Dynamically update php-fpm to use the new user and group
+RUN sed -i "s/user = www-data/user = www/g" /usr/local/etc/php-fpm.d/www.conf && \
+    sed -i "s/group = www-data/group = $group_name/g" /usr/local/etc/php-fpm.d/www.conf
+
+
+# Set the working directory
+WORKDIR /var/www
+
+# Copy the entrypoint script
+COPY ./docker/php-fpm/entrypoint.sh /usr/local/bin/entrypoint.sh
+RUN chmod +x /usr/local/bin/entrypoint.sh
+
+# Change the default command to run the entrypoint script
+ENTRYPOINT ["/usr/local/bin/entrypoint.sh"]
+
+# Expose port 9000 and start php-fpm server
+EXPOSE 9000
+CMD ["php-fpm"]
+```
+
+This Dockerfile installs the necessary PHP extensions required by Laravel, including database drivers and the Xdebug extension for debugging.
+
+### Create a Dockerfile for Workspace
+
+The workspace container is used to run Artisan commands, Composer, and NPM. Here's the Dockerfile for the workspace:
+
+```dockerfile
+# docker/workspace/Dockerfile
+# Use the official PHP CLI image as the base
+FROM php:8.3-cli
+
+# Set environment variables for user and group ID
+ARG UID=1000
+ARG GID=1000
+ARG NODE_VERSION=22.0.0
+
+# Install system dependencies and build libraries
+RUN apt-get update && apt-get install -y --no-install-recommends \
+    curl \
+    unzip \
+    libpq-dev \
+    libonig-dev \
+    libssl-dev \
+    libxml2-dev \
+    libcurl4-openssl-dev \
+    libicu-dev \
+    libzip-dev \
+    && docker-php-ext-install -j$(nproc) \
+    pdo_mysql \
+    pdo_pgsql \
+    pgsql \
+    opcache \
+    intl \
+    zip \
+    bcmath \
+    soap \
+    && pecl install redis xdebug \
+    && docker-php-ext-enable redis xdebug\
+    && curl -sS https://getcomposer.org/installer | php -- --install-dir=/usr/local/bin --filename=composer \
+    && apt-get autoremove -y && apt-get clean && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
```

Comment:
I realize these two Dockerfiles use different base images, but this really seems like it should be a multi-stage build with two targets, rather than two entirely different Dockerfiles. I'm not sure what the purpose of the workspace container is, but if it is necessary, is there some way you can use multi-stage builds to deduplicate things?

Reply:
This would be a good point of optimization, but though it looks not optimized and fpm image also can be used for cli in many cases, FPM uses "STOPSIGNAL SIGQUIT". I suppose it would be better to use the optimized version for the CLI to avoid possible side effects.